scripts/activeWindowHook.ahk

```diff
@@ -0,0 +1,307 @@
+		{
+			Obj := WinHook.Shell.Hooks
+			For key, Hook in WinHook.Shell.Hooks
+				Display .= key "|" Hook.Event "|" Hook.Func.Name "|" Hook.Title "|" Hook.Class "|" Hook.Exe "`n"
+			return Trim(Display, "`n")
+		}
+		Deregister()
+		{
+			DllCall("DeregisterShellHookWindow", UInt, A_ScriptHwnd)
+			WinHook.Shell.Hooks := "", WinHook.Shell.Events := ""		
+		}
+		Message(Event, Hwnd)  ; Private Method
+		{
+			DetectHiddenWindows, On
+			If (WinHook.Shell.Events[Event] or WinHook.Shell.Events[0]) 
+			{
+				
+				WinGetTitle, wTitle, ahk_id %Hwnd%
+				WinGetClass, wClass, ahk_id %Hwnd%
+				WinGet, wExe, ProcessName, ahk_id %Hwnd%
+				for key, Hook in WinHook.Shell.Hooks
+					if ((Hook.Title = wTitle or Hook.Title = "") and (Hook.Class = wClass or Hook.Class = "") and (Hook.Exe = wExe or Hook.Exe = "") and (Hook.Event = Event or Hook.Event = 0))
+						return Hook.Func.Call(Hwnd, wTitle, wClass, wExe, Event)
+			}
+		}
+	}
+	class Event
+	{
+		Add(eventMin, eventMax, eventProc, idProcess := 0, WinTitle := "")
+		{
+			if !WinHook.Event.Hooks
+			{
+				WinHook.Event.Hooks := {}
+				static CB_WinEventProc := RegisterCallback(WinHook.Event.Message)
+				OnExit(ObjBindMethod(WinHook.Event, "UnHookAll"))
+			}
+			hWinEventHook := DllCall("SetWinEventHook"
+				, "UInt",	eventMin						;  UINT eventMin
+				, "UInt",	eventMax						;  UINT eventMax
+				, "Ptr" ,	0x0								;  HMODULE hmodWinEventProc
+				, "Ptr" ,	CB_WinEventProc					;  WINEVENTPROC lpfnWinEventProc
+				, "UInt" ,	idProcess						;  DWORD idProcess
+				, "UInt",	0x0								;  DWORD idThread
+				, "UInt",	0x0|0x2)  						;  UINT dwflags, OutOfContext|SkipOwnProcess
+			if !IsObject(eventProc)
+				eventProc := Func(eventProc)
+			WinHook.Event.Hooks[hWinEventHook] := {eventMin: eventMin, eventMax: eventMax, eventProc: eventProc, idProcess: idProcess, WinTitle: WinTitle}
+			return hWinEventHook
+		}
+		Report(ByRef Obj:="")
+		{
+			Obj := WinHook.Event.Hooks
+			For hWinEventHook, Hook in WinHook.Event.Hooks
+				Display .= hWinEventHook "|" Hook.eventMin "|" Hook.eventMax "|" Hook.eventProc.Name "|" Hook.idProcess "|" Hook.WinTitle "`n"
+			return Trim(Display, "`n")
+		}
+		UnHook(hWinEventHook)
+		{
+				DllCall("UnhookWinEvent", "Ptr", hWinEventHook)
+				WinHook.Event.Hooks.Delete(hWinEventHook)
+		}
+		UnHookAll()
+		{
+			for hWinEventHook, Hook in WinHook.Event.Hooks
+				DllCall("UnhookWinEvent", "Ptr", hWinEventHook)
+			WinHook.Event.Hooks := "", CB_WinEventProc := ""
+		}
+		Message(event, hwnd, idObject, idChild, dwEventThread, dwmsEventTime)  ; 'Private Method 
+		{
+			DetectHiddenWindows, On
+			Hook := WinHook.Event.Hooks[hWinEventHook := this] ; this' is hidden param1 because method is called as func
+			WinGet, List, List, % Hook.WinTitle
+			Loop % List
+				if  (List%A_Index% = hwnd)
+					return Hook.eventProc.Call(hWinEventHook, event, hwnd, idObject, idChild, dwEventThread, dwmsEventTime)
+		}
+	}
+}
+
+WinHook.Shell.Add("Activated",,,,32772)
```

Comment:
This adds a hook to every change to an active window. Strictly speaking, hooks could/should be limited to windows where the active status changing is relevant to the MacroPad.